File: Introduction/9__sum_odd_elements_list.hs

-- https://www.hackerrank.com/challenges/fp-sum-of-odd-elements

-- Does not work
f1 :: [Int] -> Int
f1 arr = sum (filter (odd arr))


f2 :: [Int] -> Int
f2 arr = sum[x | x <- arr, x `mod` 2 /= 0]


f :: [Int] -> Int
f []        = 0
f (x:xs)    = odd1 x + f xs
            where odd1 n    | odd n     = n
                            | otherwise = 0

File: Introduction/9__sum_odd_elements_list.scala

// https://www.hackerrank.com/challenges/fp-sum-of-odd-elements

def f(l: List[Int]): Int = l match {
    case h :: tail => odd1(h) + f(tail)
    case Nil => 0
}

def odd1(x: Int): Int = {
    if(x % 2 != 0) x
    else 0
}


def f1(l: List[Int]): Int = l.filter(_ % 2 != 0).sum